api/service/prometheus/service.go

```diff
@@ -0,0 +1,122 @@
+// Package prometheus defines a service which is used for metrics collection
+// and health of a node in Harmony.
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"runtime/debug"
+	"runtime/pprof"
+	"time"
+
+	nodeconfig "github.com/harmony-one/harmony/internal/configs/node"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rs/zerolog"
+)
+
+// Service provides Prometheus metrics via the /metrics route. This route will
+// show all the metrics registered with the Prometheus DefaultRegisterer.
+type Service struct {
+	server     *http.Server
+	failStatus error
+}
+
+// Handler represents a path and handler func to serve on the same port as /metrics, /healthz, /goroutinez, etc.
+type Handler struct {
+	Path    string
+	Handler func(http.ResponseWriter, *http.Request)
+}
+
+var (
+	svc = &Service{}
+)
+
+// NewService sets up a new instance for a given address host:port.
+// An empty host will match with any IP so an address like ":19000" is perfectly acceptable.
+func NewService(config nodeconfig.PrometheusServerConfig, additionalHandlers ...Handler) {
+
+	if !config.HTTPEnabled {
+		getLogger().Info().Msg("Prometheus http server disabled...")
+		return
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	mux.HandleFunc("/goroutinez", svc.goroutinezHandler)
+
+	// Register additional handlers.
+	for _, h := range additionalHandlers {
+		mux.HandleFunc(h.Path, h.Handler)
+	}
+
+	getLogger().Info().Int("port", config.HTTPPort).
+		Str("ip", config.HTTPIp).
+		Msg("Starting Prometheus server")
+	endpoint := fmt.Sprintf("%s:%d", config.HTTPIp, config.HTTPPort)
+	svc.server = &http.Server{Addr: endpoint, Handler: mux}
+
+	svc.Start()
+}
+
+// StopService stop the Prometheus service
+func StopService() error {
+	return svc.Stop()
+}
+
+func (s *Service) goroutinezHandler(w http.ResponseWriter, _ *http.Request) {
+	stack := debug.Stack()
+	if _, err := w.Write(stack); err != nil {
+		getLogger().Error().Err(err).Msg("Failed to write goroutines stack")
+	}
+	if err := pprof.Lookup("goroutine").WriteTo(w, 2); err != nil {
+		getLogger().Error().Err(err).Msg("Failed to write pprof goroutines")
+	}
+}
+
+// Start the prometheus service.
+func (s *Service) Start() {
+	go func() {
+		// See if the port is already used.
+		conn, err := net.DialTimeout("tcp", s.server.Addr, time.Second)
```

Comment:
1. Why we do need to check availability before start a server? If the port is already in use, is it true that the error can be returned from ListenAndServe?
2. Consider refactor this into a new function called `checkServerAddrReachable`? This piece of code looks like it is doing a client job at the first sight.

Reply:
removed to make it simpler